txnsync/incoming.go

```diff
@@ -32,6 +34,68 @@ type incomingMessage struct {
 	encodedSize    int
 }
 
+// incomingMessageQueue manages the global incoming message queue across all the incoming peers.
+type incomingMessageQueue struct {
+	incomingMessages chan incomingMessage
+	enqueuedPeers    map[*Peer]struct{}
+	enqueuedPeersMu  deadlock.Mutex
+}
+
+// maxPeersCount defines the maximum number of supported peers that can have their messages waiting
+// in the incoming message queue at the same time. This number can be lower then the actual number of
+// connected peers, as it's used only for pending messages.
+const maxPeersCount = 1024
+
+// makeIncomingMessageQueue creates an incomingMessageQueue object and initializes all the internal variables.
+func makeIncomingMessageQueue() incomingMessageQueue {
```

Comment:
I do like to use pointer when I can.

However, I learned it (the hard way) that using a pointer comes with a risk - that pointer might be a `nil` value. And since I don't really want to handle this use case, I try to use plain objects instead when performance is not of an issue.

This falls exactly into that use case:
1. We initialize this data structure exactly one time.
2. The copying time isn't an issue.
3. Data structure isn't moving or being passed around.
4. All the methods that access this structure are pointer receiver methods. 

Try to think of this as if all these parameters were placed on the parent object.. but that we decided to pack these together for clarity.